File: algorithms/sorting/mergeSort.py
arr = [1,-1,5,9,0,100,45]

def mergeSort(arr, start, end):
    if start >= end: return
    
    mid = (start+end) // 2
    mergeSort(arr, start, mid)
    mergeSort(arr, mid+1, end)

    merge(arr,start, mid, end)

def merge(arr, start, mid, end):
    res = [None] * len(arr)
    
    left_start = start
    left_end = mid
    right_start = mid+1
    right_end = end
    
    k = start
    while left_start <= left_end and right_start <= right_end:
        if arr[left_start] < arr[right_start]:
            res[k] = arr[left_start]
            left_start += 1
        else:
            res[k] = arr[right_start]
            right_start += 1
        
        k += 1
    
    while left_start <= left_end:
        res[k] = arr[left_start]
        left_start += 1
        k += 1
    
    while right_start <= right_end:
        res[k] = arr[right_start]
        right_start += 1
        k += 1

    for x in range(start, end+1):
        arr[x] = res[x]
    
mergeSort(arr, 0, len(arr)-1)

print(arr)
    

File: algorithms/sorting/mergeSort.cpp
#include <bits/stdc++.h>
#define ll long long
#define ar array
#define pb push_back
#define vi vector<int>
#define vvi vector<vector<int>>
#define vii vector<int, int>
#define pii pair<int, int>

using namespace std;
const ll M = 1000000007;
const int mxN = 1e3;

void merge(int arr[], int l, int m, int r){

   int tmp[6];

   int i = l, left_end = m;
   int j = m + 1, right_end = r;
   int k = l;

   for (; i <= left_end && j <= right_end; k++){
       if (arr[i] < arr[j]){
           tmp[k] = arr[i];
           i++;
       }else{
           tmp[k] = arr[j];
           j++;
        }
   }

   while (i <= left_end){
       tmp[k] = arr[i];
       i++;
       k++;
   }

   while (j <= right_end){
       tmp[k] = arr[j];
       j++;
       k++;
   }

   for (int x = l; x < r+1; x++){
       arr[x] = tmp[x];
   }
    
}
 
// l is for left index and r is
// right index of the sub-array
// of arr to be sorted */
void mergeSort(int arr[],int l,int r){
    if(l>=r){
        return;//returns recursively
    }
    int m = (l+r-1)/2;
    mergeSort(arr,l,m);
    mergeSort(arr,m+1,r);
    merge(arr,l,m,r);
}
 
// UTILITY FUNCTIONS
// Function to print an array
void printArray(int A[], int size)
{
    for (int i = 0; i < size; i++)
        cout << A[i] << " ";
}
 
// Driver code
int main()
{
    int arr[] = { 12, 11, 13, 5, 6, 7 };
    int arr_size = sizeof(arr) / sizeof(arr[0]);
 
    cout << "Given array is \n";
    printArray(arr, arr_size);
 
    mergeSort(arr, 0, arr_size - 1);
 
    cout << "\nSorted array is \n";
    printArray(arr, arr_size);
    return 0;
}